### PR TITLE
[SPARK-17219][ML] enhanced NaN value handling in Bucketizer

### PR DESCRIPTION
**Pull request by VinceShieh**
## What changes were proposed in this pull request?

This PR is an enhancement of PR with commit ID:57dc326bd00cf0a49da971e9c573c48ae28acaa2.
NaN is a special type of value which is commonly seen as invalid. But We find that there are certain cases where NaN are also valuable, thus need special handling. We provided user when dealing NaN values with 3 options, to either reserve an extra bucket for NaN values, or remove the NaN values, or report an error, by setting handleNaN "keep", "skip", or "error"(default) respectively.

'''Before:
val bucketizer: Bucketizer = new Bucketizer()
          .setInputCol("feature")
          .setOutputCol("result")
          .setSplits(splits)
'''After:
val bucketizer: Bucketizer = new Bucketizer()
          .setInputCol("feature")
          .setOutputCol("result")
          .setSplits(splits)
          .setHandleNaN("keep")
## How was this patch tested?

Tests added in QuantileDiscretizerSuite, BucketizerSuite and DataFrameStatSuite

Signed-off-by: VinceShieh vincent.xie@intel.com

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala [hunk 1]
```diff
@@ -270,10 +270,10 @@ private[ml] trait HasFitIntercept extends Params {
 private[ml] trait HasHandleInvalid extends Params {
 
   /**
-   * Param for how to handle invalid entries. Options are skip (which will filter out rows with bad values), or error (which will throw an error). More options may be added later.
+   * Param for how to handle invalid entries. Options are skip (which will filter out rows with bad values), or error (which will throw an error), or keep (which will keep the bad values in certain way). More options may be added later.
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala]: I'm neutral on the complexity that this adds, but not against it. It gets a little funny to say "keep invalid data" but I think we discussed that on the JIRA

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 2]
```diff
@@ -73,15 +78,27 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
...
+
   @Since("2.0.0")
   override def transform(dataset: Dataset[_]): DataFrame = {
     transformSchema(dataset.schema)
-    val bucketizer = udf { feature: Double =>
-      Bucketizer.binarySearchForBuckets($(splits), feature)
+    val bucketizer: UserDefinedFunction = udf { (feature: Double, flag: String) =>
+      Bucketizer.binarySearchForBuckets($(splits), feature, flag)
+    }
+    val filteredDataset = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Doesn't this need to try to handle "error"?

```
val filteredDataSet = getHandleInvalid match {
  case "skip" => dataset.na.drop
  case "keep" => dataset
  case "error" => 
    if (...dataset contains NaN...) {
      throw new IllegalArgumentException(...)
   } else {
      dataset
   }
}
```

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 3]
```diff
@@ -128,8 +145,9 @@ object Bucketizer extends DefaultParamsReadable[Bucketizer] {
    * Binary searching in several buckets to place each data point.
    * @throws SparkException if a feature is < splits.head or > splits.last
    */
-  private[feature] def binarySearchForBuckets(splits: Array[Double], feature: Double): Double = {
-    if (feature.isNaN) {
+  private[feature] def binarySearchForBuckets
+  (splits: Array[Double], feature: Double, flag: String): Double = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Nit: I think the convention is to leave the open paren on the previous line

Doesn't this need to handle "skip" and "error"? throw an exception on NaN if "error" or ignore it if "skip"?

**Review by srowen — CHANGES_REQUESTED** [review]

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala (hunk 1 above)
Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala]: yeah, usually we treat NaN as a type of invalid value, but we know there are cases they are proved useful, so, it's a bit dilemma here, also with the API naming.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala (hunk 2 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: I don't see that the method handles `NaN` below. What binarySearch returns is undefined. One place or the other I think this has to be explicitly handled.

**Review by srowen — CHANGES_REQUESTED** [review]
I'll let @thunterdb or @jkbradley mostly review from here to see if this is what they intended

**Comment by jkbradley** [comment]
Thanks! I'll take a look.

Could you please fix the typo in the title? "enchanced" -> "enhanced"

**Comment by VinceShieh** [comment]
typo corrected. Thank you all. @srowen @jkbradley

File: docs/ml-features.md [hunk 4]
```diff
@@ -1104,9 +1104,11 @@ for more details on the API.
 `QuantileDiscretizer` takes a column with continuous features and outputs a column with binned
 categorical features. The number of bins is set by the `numBuckets` parameter. It is possible
 that the number of buckets used will be less than this value, for example, if there are too few
-distinct values of the input to create enough distinct quantiles. Note also that NaN values are
-handled specially and placed into their own bucket. For example, if 4 buckets are used, then
-non-NaN data will be put into buckets[0-3], but NaNs will be counted in a special bucket[4].
+distinct values of the input to create enough distinct quantiles. Note also that QuantileDiscretizer
```
Comment (jkbradley) [inline docs/ml-features.md]: (same as below) Is "possible that the number of buckets used will be less than this value" true?  It was true before this used Dataset.approxQuantiles, but I don't think it is true any longer.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 5]
```diff
@@ -73,15 +78,27 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
...
+  @Since("2.1.0")
+  def setHandleInvalid(value: String): this.type = set(handleInvalid, value)
+  setDefault(handleInvalid, "error")
+
   @Since("2.0.0")
   override def transform(dataset: Dataset[_]): DataFrame = {
     transformSchema(dataset.schema)
-    val bucketizer = udf { feature: Double =>
-      Bucketizer.binarySearchForBuckets($(splits), feature)
+    val bucketizer: UserDefinedFunction = udf { (feature: Double, flag: String) =>
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Also, at this point, you already know the value for flag, so you can just use it here, rather than making the UDF take an extra argument.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 6]
```diff
@@ -73,15 +78,27 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
...
+      Bucketizer.binarySearchForBuckets($(splits), feature, flag)
+    }
+    val filteredDataset = {
+      if ("skip" == getHandleInvalid) {
+        dataset.na.drop
+      } else {
+        dataset
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: I'd put ```dataset.toDF()``` here and not import existentials.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 7]
```diff
@@ -128,8 +145,11 @@ object Bucketizer extends DefaultParamsReadable[Bucketizer] {
    * Binary searching in several buckets to place each data point.
    * @throws SparkException if a feature is < splits.head or > splits.last
    */
-  private[feature] def binarySearchForBuckets(splits: Array[Double], feature: Double): Double = {
-    if (feature.isNaN) {
+  private[feature] def binarySearchForBuckets(
+                                               splits: Array[Double],
+                                               feature: Double,
+                                               flag: String): Double = {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Use a better name than "flag"  How about "keepNaNs"?
Also, please document this argument.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 8]
```diff
@@ -66,11 +67,13 @@ private[feature] trait QuantileDiscretizerBase extends Params
 /**
  * `QuantileDiscretizer` takes a column with continuous features and outputs a column with binned
  * categorical features. The number of bins can be set using the `numBuckets` parameter. It is
- * possible that the number of buckets used will be less than this value, for example, if there
- * are too few distinct values of the input to create enough distinct quantiles. Note also that
- * NaN values are handled specially and placed into their own bucket. For example, if 4 buckets
- * are used, then non-NaN data will be put into buckets(0-3), but NaNs will be counted in a special
- * bucket(4).
+ * possible that the number of buckets used will be less than this value, for example, if there are
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: Is this true?  It was true before this used Dataset.approxQuantiles, but I don't think it is true any longer.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 9]
```diff
@@ -66,11 +67,13 @@ private[feature] trait QuantileDiscretizerBase extends Params
...
+ * too few distinct values of the input to create enough distinct quantiles. Note also that
+ * QuantileDiscretizer will raise an error when it finds NaN value in the dataset, but user can
+ * also choose to either keep or remove NaN values within the dataset by calling setHandleInvalid.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: "calling setHandleInvalid" -> "setting `handleInvalid`"

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 10]
```diff
@@ -66,11 +67,13 @@ private[feature] trait QuantileDiscretizerBase extends Params
...
+ * If user chooses to keep NaN values, they will be handled specially and placed into their own
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: This documentation should go in the handleInvalid Param doc string.

File: mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala [hunk 11]
```diff
@@ -114,6 +115,7 @@ class BucketizerSuite extends SparkFunSuite with MLlibTestSparkContext with Defa
           .setInputCol("feature")
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala]: Just set splits here; there's no need to set the other Params for this test.

File: mllib/src/test/scala/org/apache/spark/ml/feature/QuantileDiscretizerSuite.scala [hunk 12]
```diff
@@ -83,10 +83,20 @@ class QuantileDiscretizerSuite
       .setOutputCol("result")
       .setNumBuckets(numBuckets)
 
-    // Reserve extra one bucket for NaN
-    val expectedNumBuckets = discretizer.fit(df).getSplits.length - 1
-    val result = discretizer.fit(df).transform(df)
-    val observedNumBuckets = result.select("result").distinct.count
+    // Reserve extra one bucket for NaN values
+    discretizer.setHandleInvalid("keep")
+    var expectedNumBuckets = discretizer.fit(df).getSplits.length - 1
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/QuantileDiscretizerSuite.scala]: I'd also recommend directly testing the values of the splits.  The current test makes sure that handleInvalid is passed to the bucketizer correctly, which is important but separate.

Also, please use vals (not vars) for clarity.  I'd recommend making a helper method for lines 87-93, which can be reused for the test of handleInvalid = "skip"

File: mllib/src/test/scala/org/apache/spark/ml/feature/QuantileDiscretizerSuite.scala [hunk 13]
```diff
@@ -83,10 +83,20 @@ class QuantileDiscretizerSuite
...
+    var result = discretizer.fit(df).transform(df)
+    var observedNumBuckets = result.select("result").distinct.count
+    assert(observedNumBuckets == expectedNumBuckets,
+      s"Observed number of buckets are not correct." +
+        s" Expected $expectedNumBuckets but found $observedNumBuckets")
+
+    // Remove NaN values
+    discretizer.setHandleInvalid("skip")
+    expectedNumBuckets = discretizer.fit(df).getSplits.length - 2
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/QuantileDiscretizerSuite.scala]: Same here; I'd also recommend directly testing the values of the splits.

File: python/pyspark/ml/feature.py [hunk 14]
```diff
@@ -1157,9 +1157,11 @@ class QuantileDiscretizer(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadab
     categorical features. The number of bins can be set using the :py:attr:`numBuckets` parameter.
     It is possible that the number of buckets used will be less than this value, for example, if
```
Comment (jkbradley) [inline python/pyspark/ml/feature.py]: Here too: no longer the case

**Review by jkbradley — COMMENTED** [review]
Done with review.  Thanks for the PR!

Other comments:
- Add unit test for `DataFrameStatFunctions.approxQuantile` to check for handling of NaN values.
- Create JIRA for adding handleInvalid to Python API.

File: docs/ml-features.md (hunk 4 above)
Comment (VinceShieh) [inline docs/ml-features.md]: In cases when the number of buckets requested by the user is greater than the number of distinct splits generated from Bucketizer, the returned number of buckets will be less than requested.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala (hunk 8 above)
Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: same as the comment above

File: python/pyspark/ml/feature.py (hunk 14 above)
Comment (VinceShieh) [inline python/pyspark/ml/feature.py]: same as the comment above.

**Comment by VinceShieh** [comment]
Thanks for your valuable suggestions. @jkbradley @srowen

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 15]
```diff
@@ -73,15 +74,51 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
...
+  def setHandleNaN(value: String): this.type = set(handleNaN, value)
+  setDefault(handleNaN, "error")
+
   @Since("2.0.0")
   override def transform(dataset: Dataset[_]): DataFrame = {
     transformSchema(dataset.schema)
-    val bucketizer = udf { feature: Double =>
-      Bucketizer.binarySearchForBuckets($(splits), feature)
+
+    val bucketizer: UserDefinedFunction = udf { (feature: Double) =>
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Ah, sorry, one more comment.  I'm not quite sure how closure capture behaves currently, but it might be good to define local vals for ```$(splits)``` and ```getHandleNaN.isDefined && getHandleNaN.get```.  Since these reference methods in the Bucketizer class, I believe the UDF may capture the whole Bucketizer class instead of just those vals.

After you define them in local vals here, you can use those vals in this UDF.

File: docs/ml-features.md (hunk 4 above)
Comment (jkbradley) [inline docs/ml-features.md]: Yep, you're right

File: mllib/src/test/scala/org/apache/spark/ml/feature/QuantileDiscretizerSuite.scala [hunk 16]
```diff
@@ -82,14 +82,23 @@ class QuantileDiscretizerSuite
...
-    val result = discretizer.fit(df).transform(df)
-    val observedNumBuckets = result.select("result").distinct.count
-    assert(observedNumBuckets == expectedNumBuckets,
-      s"Observed number of buckets are not correct." +
-        s" Expected $expectedNumBuckets but found $observedNumBuckets")
+    val splits = Array(Double.NegativeInfinity, 1.0, Double.PositiveInfinity)
+
+    for (handleNaN <- Array("keep", "skip")) {
+      discretizer.setHandleNaN(handleNaN)
+      val expectedNumBuckets = {
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/QuantileDiscretizerSuite.scala]: This is kind of complex logic just to get the expected number of splits.  I'd recommend just putting the expected bucket *values* in the original DataFrame:
```
val df = sc.parallelize(Seq(
  (1.0, /*expected value for option "keep"*/, /*expected value for option "skip"*/),
  ...
)).toDF("input", "expectedKeep", "expectedSkip")
```
Then you can compare with the actual values.  That'll be an easier test to understand IMO.

File: python/pyspark/ml/feature.py [hunk 17]
```diff
@@ -1157,9 +1157,11 @@ class QuantileDiscretizer(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadab
...
     there are too few distinct values of the input to create enough distinct quantiles. Note also
-    that NaN values are handled specially and placed into their own bucket. For example, if 4
-    buckets are used, then non-NaN data will be put into buckets(0-3), but NaNs will be counted in
-    a special bucket(4).
+    that QuantileDiscretizer will raise an error when it finds NaN value in the dataset, but user
```
Comment (jkbradley) [inline python/pyspark/ml/feature.py]: Actually no need to update Python API until it is updated to include handleNaN

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 18]
```diff
@@ -100,6 +102,24 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
...
 
+  /**
+   * Param for how to handle NaN entries. Options are skip (which will filter out rows with
+   * NaN values), or error (which will throw an error), or keep (which will make the NaN
+   * values an extra bucket). More options may be added later.
+   *
+   * @group param
+   */
+  @Since("2.1.0")
+  val handleNaN: Param[String] = new Param[String](this, "handleNaNs", "how to handle NaN" +
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: Add getHandleNaN

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 19]
```diff
@@ -100,6 +102,24 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
   @Since("1.6.0")
   def setOutputCol(value: String): this.type = set(outputCol, value)
 
+  /**
+   * Param for how to handle NaN entries. Options are skip (which will filter out rows with
+   * NaN values), or error (which will throw an error), or keep (which will make the NaN
+   * values an extra bucket). More options may be added later.
+   *
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: State default

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 20]
```diff
@@ -73,15 +74,51 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
   @Since("1.4.0")
   def setOutputCol(value: String): this.type = set(outputCol, value)
 
+  /**
+   * Param for how to handle NaN entries. Options are skip (which will filter out rows with
+   * NaN values), or error (which will throw an error), or keep (which will make the NaN
+   * values an extra bucket). More options may be added later.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Remove "More options may be added later."

Also state default.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 21]
```diff
@@ -73,15 +74,51 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
...
+   *
+   * @group param
+   */
+  @Since("2.1.0")
+  val handleNaN: Param[String] = new Param[String](this, "handleNaNs", "how to handle NaN" +
+    "entries. Options are skip (which will filter out rows with NaN values), or error" +
+    "(which will throw an error), or keep (which will make the NaN values an extra bucket)." +
+    "More options may be added later", ParamValidators.inArray(Array("skip", "error", "keep")))
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Remove "More options may be added later."

**Review by jkbradley — COMMENTED** [review]
Thanks for the updates!  A few more comments.

Also, I see you renamed the param to handleNaN.  Do you think we will ever want to handle null values too?  It might be good to stick with "handleInvalid" in case we add null support.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 22]
```diff
@@ -73,15 +74,52 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
   @Since("1.4.0")
   def setOutputCol(value: String): this.type = set(outputCol, value)
 
+  /**
+   * Param for how to handle invalid entries. Options are skip (which will filter out rows with
+   * invalid values), or error (which will throw an error), or keep (which will keep the invalid
+   * values in certain way). Default behaviour is to report an error for invalid entries.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: I'd just write: ```default: "error"```
Rewording as "report" instead of "throw" could confuse people.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 23]
```diff
@@ -73,15 +74,52 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
...
+  @Since("2.1.0")
+  val handleInvalid: Param[String] = new Param[String](this, "handleInvalid", "how to handle" +
+    "invalid entries. Options are skip (which will filter out rows with invalid values), or" +
+    "error (which will throw an error), or keep (which will keep the invalid values" +
+    " in certain way). Default behaviour is to report an error for invalid entries.",
+    ParamValidators.inArray(Array("skip", "error", "keep")))
+
+  /** @group getParam */
+  @Since("2.1.0")
+  def gethandleInvalid: Option[Boolean] = $(handleInvalid) match {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: This should just return ```$(handleInvalid)```, just like any other Param getter method.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 24]
```diff
@@ -100,6 +102,33 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
   @Since("1.6.0")
   def setOutputCol(value: String): this.type = set(outputCol, value)
 
+  /**
+   * Param for how to handle invalid entries. Options are skip (which will filter out rows with
+   * invalid values), or error (which will throw an error), or keep (which will keep the invalid
+   * values in certain way). Default behaviour is to report an error for invalid entries.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: I'd just write: ```default: "error"```
Rewording as "report" instead of "throw" could confuse people.

File: python/pyspark/ml/feature.py (hunk 17 above)
Comment (jkbradley) [inline python/pyspark/ml/feature.py]: This isn't available in Python yet, so can you please revert this change to feature.py?

**Review by jkbradley — COMMENTED** [review]
Thanks for the updates.
Also, when you push further updates, could you please just push a new commit?  Rebasing makes it harder to review since it makes it impossible to connect your most recent changes with the recent reviewer comments.  Thanks!

**Comment by jkbradley** [comment]
Thanks, you still need to remove the change in feature.py, but other than that, this should be ready.

**Comment by VinceShieh** [comment]
sorry, I must have forgotten to commit the changes.
All done now. Thanks for reviewing.

**Comment by jkbradley** [comment]
Thanks for the update!  On a last glance through, I spotted a few things to fix up.  Since some had to do with wording, I thought it'd be easier just to send a PR to your PR.  Can you please review this and merge it if it looks Ok to you?  Thanks!  Here it is: [https://github.com/VinceShieh/spark/pull/2]

**Comment by jkbradley** [comment]
Thanks for merging that!  This LGTM
Merging with master
Thanks a lot @VinceShieh !
